File: api/hosts.go
package api

import (
	"encoding/json"
	"net/http"

	real_debrid "github.com/sushydev/real_debrid_go"
)

type HostBasic struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Image string `json:"image"`
}

func GetHosts(client *real_debrid.Client) (map[string]HostBasic, error) {
	url := client.GetUrl("/hosts")
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, err
	}
	out := map[string]HostBasic{}
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}

type HostStatus struct {
	ID                string `json:"id"`
	Name              string `json:"name"`
	Image             string `json:"image"`
	Supported         int    `json:"supported"`
	Status            string `json:"status"`
	CheckTime         string `json:"check_time"`
	CompetitorsStatus map[string]struct {
		Status    string `json:"status"`
		CheckTime string `json:"check_time"`
	} `json:"competitors_status"`
}

func GetHostsStatus(client *real_debrid.Client) (map[string]HostStatus, error) {
	url := client.GetUrl("/hosts/status")
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, err
	}
	out := map[string]HostStatus{}
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}

func GetHostsRegex(client *real_debrid.Client) ([]string, error) {
	url := client.GetUrl("/hosts/regex")
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, err
	}
	var out []string
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}

func GetHostsRegexFolder(client *real_debrid.Client) ([]string, error) {
	url := client.GetUrl("/hosts/regexFolder")
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, err
	}
	var out []string
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}

func GetHostsDomains(client *real_debrid.Client) ([]string, error) {
	url := client.GetUrl("/hosts/domains")
	req, err := http.NewRequest("GET", url.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := client.HandleResponseCode(resp, 200); err != nil {
		return nil, err
	}
	var out []string
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	return out, nil
}

File: lib/real_debrid/api/hosts.ex
defmodule RealDebrid.Api.Hosts do
  @moduledoc """
  Hosts-related API functions.
  """

  alias RealDebrid.Client

  @type host_basic :: %{
          id: String.t(),
          name: String.t(),
          image: String.t()
        }

  @type host_status :: %{
          id: String.t(),
          name: String.t(),
          image: String.t(),
          supported: integer(),
          status: String.t(),
          check_time: String.t(),
          competitors_status: map()
        }

  @doc """
  Get list of supported hosts.

  ## Returns

    - `{:ok, hosts}` - Map of host ID to host info
    - `{:error, reason}` on failure
  """
  @spec get(Client.t()) :: {:ok, map()} | {:error, term()}
  def get(%Client{} = client) do
    case Client.get(client, "/hosts") do
      {:ok, body, _headers} ->
        hosts =
          body
          |> Enum.map(fn {key, value} ->
            {key,
             %{
               id: value["id"],
               name: value["name"],
               image: value["image"]
             }}
          end)
          |> Map.new()

        {:ok, hosts}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Get status of supported hosts.

  ## Returns

    - `{:ok, hosts}` - Map of host ID to host status
    - `{:error, reason}` on failure
  """
  @spec get_status(Client.t()) :: {:ok, map()} | {:error, term()}
  def get_status(%Client{} = client) do
    case Client.get(client, "/hosts/status") do
      {:ok, body, _headers} ->
        hosts =
          body
          |> Enum.map(fn {key, value} ->
            {key,
             %{
               id: value["id"],
               name: value["name"],
               image: value["image"],
               supported: value["supported"],
               status: value["status"],
               check_time: value["check_time"],
               competitors_status: value["competitors_status"] || %{}
             }}
          end)
          |> Map.new()

        {:ok, hosts}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Get supported hosts regex patterns.

  ## Returns

    - `{:ok, patterns}` - List of regex patterns
    - `{:error, reason}` on failure
  """
  @spec get_regex(Client.t()) :: {:ok, [String.t()]} | {:error, term()}
  def get_regex(%Client{} = client) do
    case Client.get(client, "/hosts/regex") do
      {:ok, body, _headers} -> {:ok, body}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Get supported hosts folder regex patterns.

  ## Returns

    - `{:ok, patterns}` - List of folder regex patterns
    - `{:error, reason}` on failure
  """
  @spec get_regex_folder(Client.t()) :: {:ok, [String.t()]} | {:error, term()}
  def get_regex_folder(%Client{} = client) do
    case Client.get(client, "/hosts/regexFolder") do
      {:ok, body, _headers} -> {:ok, body}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc """
  Get supported hosts domains.

  ## Returns

    - `{:ok, domains}` - List of supported domains
    - `{:error, reason}` on failure
  """
  @spec get_domains(Client.t()) :: {:ok, [String.t()]} | {:error, term()}
  def get_domains(%Client{} = client) do
    case Client.get(client, "/hosts/domains") do
      {:ok, body, _headers} -> {:ok, body}
      {:error, reason} -> {:error, reason}
    end
  end
end
